extracters/cleaners.py

```diff
@@ -22,26 +22,50 @@ def share_parse(value):
 
 def ymd_to_datetime(year, month, day, which="begin"):
 
-	# These might raise ValueError if the string is not an int
-	# which is fine at this stage
+
 	if type(year) is not int:
-		year = int(year)
+		try:
+			year = int(year)
+		except:
+			print("DATE CLEAN: year is %r" % year)
+			year = None
+
 	if type(month) is not int:
-		month = int(month)
+		try:
+			month = int(month)
+		except:
+			print("DATE CLEAN: month is %r" % month)
+			month = None
+
 	if type(day) is not int:
-		day = int(day)
+		try:
+			day = int(day)
+		except:
+			day = None
+
+	if not year:
+		if which == "begin":
+			year = 0
+		else:
+			year = 9999
 
-	if not month:
+	if not month or month > 12 or month < 0:
 		if which == "begin":
 			month = 1
 		else:
 			month = 12
-	if not day:
+
+	try:
+		maxday = calendar.monthrange(year, month)[1]
+	except:
+		maxday = 31
+		
+	if not day or day > maxday or day < 0:
```

Comment:
The logic here feels strange, so just want to confirm that these cases make sense:

* If `which == "begin"` and `day > 31`, then day gets reset to `1`
* if `which == "end"` and `day < 0`, day gets set to `maxday`

Reply:
The cases I'm thinking of are when some random value gets stuffed into the field ... human error, basically.  E.g. the 150 foot long painting, when the price ended up in width.

So agree that day = 31 in a month that has 30 days is probably just a typo and they really meant 30 ... but knowing when to pick that rather than day = 500 when the code wants the earliest possible day, I think we'll generate less impossible situations by being defensive and pessimistic in cleaning code.